libs/ballot-interpreter-nh-next/src/timing_marks.rs

```diff
@@ -0,0 +1,1038 @@
+        self.0.eq(&other.0)
+    }
+}
+
+impl PartialOrd for OvalMarkScore {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+#[derive(Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ScoredOvalMark {
+    /// The location of the oval mark in the grid. Uses side/column/row, not
+    /// x/y.
+    pub location: GridLocation,
+
+    /// The score for the match between the source image and the template. This
+    /// is the highest value found when looking around `expected_bounds` for the
+    /// oval. 100% is a perfect match.
+    pub match_score: OvalMarkScore,
+
+    /// The score for the fill of the oval at `matched_bounds`. 100% is
+    /// perfectly filled.
+    pub fill_score: OvalMarkScore,
+
+    /// The expected bounds of the oval mark in the scanned source image.
+    pub expected_bounds: Rect,
+
+    /// The bounds of the oval mark in the scanned source image that was
+    /// determined to be the best match.
+    pub matched_bounds: Rect,
+
+    /// The cropped source image at `matched_bounds`.
+    #[serde(skip_serializing)]
+    pub source_image: GrayImage,
+
+    /// The cropped source image at `matched_bounds` with each pixel binarized
+    /// to either 0 (black) or 255 (white).
+    #[serde(skip_serializing)]
+    pub binarized_source_image: GrayImage,
+
+    /// A binarized diff image of `binarized_source_image` with the template.
+    /// The more white pixels, the better the match.
+    #[serde(skip_serializing)]
+    pub match_diff_image: GrayImage,
+
+    /// A binarized diff image of `binarized_source_image` with the fill of the
+    /// template. The more black pixels, the better the fill.
+    #[serde(skip_serializing)]
+    pub fill_diff_image: GrayImage,
+}
+
+impl std::fmt::Debug for ScoredOvalMark {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "ScoredOvalMark {{ location: {:?}, match_score: {}, fill_score: {}, matched_bounds: {:?} }}",
+            self.location, self.match_score, self.fill_score, self.matched_bounds
+        )
+    }
+}
+
+pub const DEFAULT_MAXIMUM_SEARCH_DISTANCE: u32 = 7;
+
+pub type ScoredOvalMarks = Vec<(GridPosition, Option<ScoredOvalMark>)>;
+
+#[time]
+pub fn score_oval_marks_from_grid_layout(
+    img: &GrayImage,
+    oval_template: &GrayImage,
+    timing_mark_grid: &TimingMarkGrid,
+    grid_layout: &GridLayout,
+    side: BallotSide,
+    debug: &ImageDebugWriter,
+) -> ScoredOvalMarks {
+    let threshold = otsu_level(img);
+
+    let scored_ovals = &grid_layout
+        .grid_positions
+        .par_iter()
+        .flat_map(|grid_position| {
+            let location = grid_position.location();
+
+            if location.side != side {
+                return vec![];
+            }
+
+            timing_mark_grid
+                .point_for_location(location.column, location.row)
+                .map_or_else(
```

Comment:
Wow I got so confused at first that the default value comes first in `map_or_else`

Reply:
Yeah I didn't find that intuitive either.